python/mage/constraint_programming/vrp_cp_solver.py

```diff
@@ -0,0 +1,353 @@
+
+    def _initialize(self):
+        for node_index in range(len(self.distance_matrix)):
+            if node_index in self._location_node_ids:
+                self._initialize_location_node(node_index)
+
+    def _initialize_location_node(self, node_index: int):
+        self._time_vars[node_index] = self._model.Var(value=0, lb=0, integer=False)
+
+        # Initialize starting point and sinking point for every vehicle
+        self._add_variable((self.SOURCE_INDEX, node_index))
+        self._add_variable((node_index, self.SINK_INDEX))
+
+        # For every node, draw lengths from and to it, with duration of edges
+        out_vars = self._add_adjacent_output_edge_variables(node_index)
+        in_vars = self._add_adjacent_input_edge_variables(node_index)
+
+        # Either it was a beginning node, or a vehicle has visited it in the drive.
+        if len(out_vars) > 0:
+            self._model.Equation(
+                self._edge_chosen_vars[(node_index, self.SINK_INDEX)] + sum(out_vars)
+                == 1
+            )
+
+        if len(in_vars) > 0:
+            self._model.Equation(
+                self._edge_chosen_vars[(self.SOURCE_INDEX, node_index)] + sum(in_vars)
+                == 1
+            )
+
+    def _add_adjacent_output_edge_variables(
+        self, node_index: int
+    ) -> List[Tuple[int, int]]:
+        edges_vars = []
+
+        for adjacent_node in range(len(self.distance_matrix)):
+            if adjacent_node == self.depot_index:
+                continue
+
+            edge = (node_index, adjacent_node)
+            var = self._add_variable(edge)
+            edges_vars.append(var)
+
+        return edges_vars
+
+    def _add_adjacent_input_edge_variables(
+        self, node_index: int
+    ) -> List[Tuple[int, int]]:
+        edges_vars = []
+
+        for adjacent_node in range(len(self.distance_matrix)):
+            if adjacent_node == self.depot_index:
+                continue
+
+            edge = (adjacent_node, node_index)
+            var = self._add_variable(edge)
+            edges_vars.append(var)
+
+        return edges_vars
+
+    def _add_variable(self, edge: Tuple[int, int]) -> GEKKO.Var:
+        var = self._edge_chosen_vars.get(edge)
+
+        if var is None:
+            var = self._model.Var(value=0, lb=0, ub=1, integer=True)
+            self._edge_chosen_vars[edge] = var
+
+        return var
+
+    def _add_constraints(self):
+        """
+        Add global constraints to the solver.
+        """
+        for constraint in self._constraints:
+            constraint.apply_constraint()
+
+    def _add_objective(self):
+        intermedias_sum = 0
+        for edge, variable in self._edge_chosen_vars.items():
+            duration = self.get_distance(edge)
+            intermedias_sum += self._model.Intermediate(duration * variable)
+
+        self._model.Obj(intermedias_sum)
+
+    def _add_options(self):
+        self._model.options.SOLVER = 1
```

Comment:
Add documentation for this solver, it is uncertain what you mean by `SOLVER = 1`